### ISSUE TITLE
[BUG] Entry points duplicated when package is installed editable

### ISSUE THREAD

**Issue by freakboy3742** [pypa/setuptools#3649]
### setuptools version

65.5.0, possibly earlier

### Python version

Py3.10+; or any project using the importlib_metadata>=4.4 compatibility shim

### OS

All

### Description

If a package declares an entry point in `setup.cfg` *and* the package has a `pyproject.toml` file, and the project is installed as editable, the entry point is duplicated.

This does not occur if the project does *not* have a `pyproject.toml`, or if the project is installed non-editable (either from source, or from a wheel).

### Expected behavior

A single entry point object should be returned for every registered entry point.

### How to Reproduce

Define a project with an entry point in `setup.cfg`
```
[options.entry_points]
myproject.plugin =
    cheese = cheese_grater
```
and an empty `pyproject.toml`.

Install that package as editable (`pip install -e path/to/project`), then retrieve the entry points with `from importlib.metadata import entry_points; entry_points(group="myproject.plugin")`). The returned list of EntryPoints appears will contain duplicate entries.

You only get a single entry point if:
* You create a wheel and install the wheel instead of installing as editable; OR
* You install from source, but *non* editable (i.e., `pip install path/to/project`); OR
* You delete the pyproject.toml and install as editable.

The problem appears to be caused because PEP517 installation mode causes two sets of metadata to be generated for each project being installed.

If `pyproject.toml` is present, PEP517 mode is enabled, which causes an `__editable__.myproject-1.2.3.pth` and `myproject-1.2.3.dist-info` file to be produced in site-packages. However, installing using setuptools causes a `.egg-info` file to be produced in the project source file. The `.pth` file in site-packages links to the source folder that contains the `.egg-info` file. The `.egg-info` file duplicates the information in the `.dist-info file`, causing a duplicated registration of the entry point.

If `pyproject.toml` is *not* present, a `myproject.egg-link` file is produced in site-packages, and therefore only the metadata in the `.egg-info` file is used.

Two workarounds:
* Manually delete the `.egg-info` file produced in the project folder as a result of installing as editable
* Use `set()` to remove duplicates from the list of entry points. The duplicate entry point objects evaluate as equal, so `set()` will remove the duplicated entries.

### Output

```
>>> from importlib.metadata import entry_points
>>> entry_points(group="my project.plugin")
[EntryPoint(name="cheese", value="cheese_grater", group="myproject.plugin"), EntryPoint(name="cheese", value="cheese_grater", group="myproject.plugin")]
```

**Comment by abravalheri** [comment]
Hi @freakboy3742, you already got the gist of what is happening: `importlib-metadata` takes a design decision to consider both `.egg-info` and `.dist-info` inside all entries in `sys.path`.

Projects with flat-layouts will have an `.egg-info` directory created in the project root.

I think the problem here is the assumption that there will only be one `.dist-info`/`.egg-info` directory in `sys.path` for a given distribution name (and therefore a single set of entry-points). If I am not wrong this is not a guarantee `importlib-metadata` gives to their users.

In the future we would like to remove `.egg-info` from setuptools pipeline, but for the time being we are not focused on that.

I also think that we cannot classify this as a bug in setuptools. The existence of `.egg-info` directory predates tools like `importlib-metadata`.

**Comment by benoit-pierre** [comment]
That does not make any sense: that would mean you can't have 2 different versions of the same package, one in the system site, and another in your user site.

See https://github.com/python/importlib_metadata/issues/410 for a similar issue. Which exact version of `importlib_metadata` are you using, @freakboy3742 ?

**Comment by abravalheri** [comment]
> That does not make any sense: that would mean you can't have 2 different versions of the same package, one in the system site, and another in your user site.

Hi @benoit-pierre, could you please elaborate a bit more what does not make sense? In my previous comment I what I said is:

> I think the problem here is the assumption that there will only be one .dist-info/.egg-info directory in sys.path for a given distribution name

This assumption that I am referring here is from the user part. I agree that you should be able to have 2 different versions of the same package installed in two different locations. I don't think my previous statement contradicts that...

**Comment by benoit-pierre** [comment]
You saying that `importlib_metadata` does not guarantee that only one set of entry points will be returned, no?

**Comment by abravalheri** [comment]
> You saying that `importlib_metadata` does not guarantee that only one set of entry points will be returned, no?

What I am sure is that `importlib_metadata` does not guarantee that only one distribution is found for a given name. For example:


```bash
virtualenv -p python3.10 /tmp/experiment
/tmp/experiment/bin/python -m pip install -U pip importlib_metadata
/tmp/experiment/bin/python -m pip install 'flit==3.7.1'
mkdir /tmp/other-entry-to-sys-path
cp -r /tmp/experiment/lib/python3.10/site-packages/flit-3.7.1.dist-info/ /tmp/other-entry-to-sys-path/flit.egg-info/
cd /tmp/other-entry-to-sys-path

/tmp/experiment/bin/python -c 'import importlib_metadata; print(list(importlib_metadata.distributions(name="flit")))'
# [<importlib_metadata.PathDistribution object at 0x7f613ed3e860>, <importlib_metadata.PathDistribution object at 0x7f613ed3f0a0>]

```

Then I assumed this might be causing the problem with duplicated entry-points. I might be wrong here, but we can see this behaviour doing the following:

```bash
/tmp/experiment/bin/python -c 'import importlib.metadata; print("\n".join(map(str, importlib.metadata.entry_points(group="console_scripts"))))'
# EntryPoint(name='flit', value='flit:main', group='console_scripts')
# EntryPoint(name='normalizer', value='charset_normalizer.cli.normalizer:cli_detect', group='console_scripts')
# EntryPoint(name='docutils', value='docutils.__main__:main', group='console_scripts')
# EntryPoint(name='flit', value='flit:main', group='console_scripts')
# EntryPoint(name='pip', value='pip._internal.cli.main:main', group='console_scripts')
# EntryPoint(name='pip3', value='pip._internal.cli.main:main', group='console_scripts')
# EntryPoint(name='pip3.10', value='pip._internal.cli.main:main', group='console_scripts')
# EntryPoint(name='wheel', value='wheel.cli:main', group='console_scripts')
```

**Comment by freakboy3742** [comment]
@benoit-pierre I'm only using `importlib_metadata` on Python 3.8 and Python 3.9; I can confirm it appears that the problem was fixed in 4.11.4. On Python 3.10 and 3.11, I'm using the stdlib version of importlib.metadata, the problem is present in both versions.
